pylgbst/comms/cbluepy.py

```diff
@@ -0,0 +1,123 @@
+import re
+import logging
+from threading import Thread
+import time
+from contextlib import contextmanager
+from enum import Enum
+
+from bluepy import btle
+
+from pylgbst.comms import Connection, LEGO_MOVE_HUB
+from pylgbst.constants import MOVE_HUB_HW_UUID_CHAR
+from pylgbst.utilities import str2hex, queue
+
+log = logging.getLogger('comms-bluepy')
+
+COMPLETE_LOCAL_NAME_ADTYPE = 9
+KILL_ON_DISPATCHER_EXCEPTION = False
+
+
+def _get_iface_number(controller):
+    """bluepy uses iface numbers instead of full names."""
+    if not controller:
+        return None
+    m = re.search(r'hci(\d+)$', controller)
+    if not m:
+        raise ValueError('Cannot find iface number in {}.'.format(controller))
+    return int(m.group(1))
+
+
+class BluepyDelegate(btle.DefaultDelegate):
+    def __init__(self, handler):
+        btle.DefaultDelegate.__init__(self)
+
+        self._handler = handler
+
+    def handleNotification(self, cHandle, data): 
+        log.debug('Incoming notification')
+        self._handler(cHandle, data)
+
+
+# We need a separate thread to wait for notifications, 
+# but calling peripheral's methods from different threads creates issues,
+# so we will wrap all the calls into a thread
+class BluepyThreadedPeripheral(object):
+    def __init__(self, addr, addrType, controller):
+        self._call_queue = queue.Queue()
+        self._addr = addr
+        self._addrType = addrType
+        self._iface_number = _get_iface_number(controller)
+                
+        self._dispatcher_thread = Thread(target=self._dispatch_calls)
+        self._dispatcher_thread.setDaemon(True)
+        self._dispatcher_thread.setName("Bluepy call dispatcher")
+        self._dispatcher_thread.start()
+
+    def _dispatch_calls(self):
+        self._peripheral = btle.Peripheral(self._addr, self._addrType, self._iface_number)
+
+        while True:
+            try:
+                try:
+                    method = self._call_queue.get(False)
+                    method()
+                except queue.Empty:
+                    pass
+                self._peripheral.waitForNotifications(1.)
+            except Exception as ex:
+                log.exception('Exception in call dispatcher thread', exc_info=ex)
+                if KILL_ON_DISPATCHER_EXCEPTION:
+                    exit(1)
```

Comment:
This is very bad practive. Code should throw exception and it's responsibility of code caller to decide what to do with the fact.

Reply:
The thing was introduced purely as a (slippery) way to test implementation of the call dispatcher. What are your thoughts on how to do the async testing properly here?

Reply:
It's simple, just let the exception go. I see no graceful shutdown code, so you can just let Python interrupt the thread.

Another approach is to use boolean variable in `while` condition and change its value in case exception occurs.

Reply:
But i don't want the exception to interrupt the thread in production. It is better to just report it in this case. The purpose of the variable is only for test to fail in case an incorrect method is called on peripheral, but still make sure that the methods are actually asynchronously called.

Reply:
Then, your thread has to be interruptible from outside. You have boolean field in the class and do `while` on that field. Once your `disconnect()` method is called, you set boolean field to false and thread stops iterating.

Reply:
Ok! Please see the update and let me know your thoughts on the new implementation.